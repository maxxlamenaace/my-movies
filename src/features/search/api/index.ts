import { search } from './search/search';

export { search };
