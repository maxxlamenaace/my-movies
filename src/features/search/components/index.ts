import SearchBar from './search-bar/SearchBar';

export { SearchBar };
