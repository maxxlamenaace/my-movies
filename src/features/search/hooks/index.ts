import { useSearch } from './use-search/use-search';

export { useSearch };
