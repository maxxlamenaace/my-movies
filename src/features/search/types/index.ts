import { Actor } from '@/features/actors';
import { Media } from '@/features/medias';

export type SearchType = 'movie' | 'serie' | 'actor';
export type SearchResponse = (Media | Actor)[];
