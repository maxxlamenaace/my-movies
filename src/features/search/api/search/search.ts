import API from '@/api/client';

import { SearchResponse, SearchType } from '@/features/search';

export const search = async <T extends SearchType>({
  type,
  query,
  page = 1,
}: {
  type: T;
  query: string;
  page?: number;
}): Promise<SearchResponse> => {
  const { results } = await API.get<Record<string, never>, { results: SearchResponse }>(
    `/search/${type === 'actor' ? 'person' : type === 'serie' ? 'tv' : type}`,
    { params: { query, page } },
  );

  return results;
};
