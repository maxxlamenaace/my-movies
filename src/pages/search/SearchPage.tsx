import React, { useEffect, useState } from 'react';

import { LoadingButton } from '@mui/lab';
import { Box, Toolbar, Typography } from '@mui/material';

import { SearchType, useSearch, SearchBar } from '@/features/search';
import { MediasGrid, Media } from '@/features/medias';
import { Actor, ActorsGrid } from '@/features/actors';
import { globalStyles } from '@/theme/styles';
import { Wrapper } from '@/features/common';
import { useAppStore } from '@/stores';

const SearchPage: React.FC = () => {
  const [query, setQuery] = useState('');
  const [searchType, setSearchType] = useState<SearchType>('movie');
  const [currentPage, setCurrentPage] = useState(1);

  const { setLoading } = useAppStore();
  const { loading, results } = useSearch({ query, type: searchType, currentPage });

  useEffect(() => {
    if (!loading) {
      setLoading(false);
    }
  }, [loading]);

  const onLoadMore = () => {
    setCurrentPage(currentPage + 1);
  };

  return (
    <Wrapper>
      <Toolbar />
      <Box sx={{ ...globalStyles.container }}>
        <SearchBar onChange={setQuery} onSearchTypeChange={setSearchType} searchType={searchType} />

        {results.length > 0 ? (
          searchType !== 'actor' ? (
            <>
              <MediasGrid
                mediaType={searchType === 'serie' ? 'tv' : searchType}
                medias={results as Media[]}
              />
              <LoadingButton
                sx={{ marginTop: 5 }}
                fullWidth
                color='primary'
                variant='contained'
                loading={loading}
                onClick={onLoadMore}
              >
                Load more
              </LoadingButton>
            </>
          ) : (
            <>
              <ActorsGrid actors={results as Actor[]} />
              <LoadingButton
                sx={{ marginTop: 5 }}
                fullWidth
                color='primary'
                variant='contained'
                loading={loading}
                onClick={onLoadMore}
              >
                Load more
              </LoadingButton>
            </>
          )
        ) : (
          query && !loading && <Typography>No results found</Typography>
        )}
      </Box>
    </Wrapper>
  );
};

export default SearchPage;
