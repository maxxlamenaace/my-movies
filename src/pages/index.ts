import HomePage from './home/HomePage';
import ActorDetailsPage from './actor-details/ActorDetailsPage';
import SearchPage from './search/SearchPage';
import MediasListPage from './medias-list/MediasListPage';
import MediaDetailsPage from './media-details/MediaDetailsPage';
import FavoritesPage from './favorites/FavoritesPage';

export { HomePage, ActorDetailsPage, SearchPage, MediasListPage, MediaDetailsPage, FavoritesPage };
