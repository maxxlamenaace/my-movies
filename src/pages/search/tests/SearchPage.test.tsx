jest.mock('@/features/search/hooks', () => ({
  ...jest.requireActual('@/features/search/hooks'),
  useSearch: jest.fn(),
}));

import React from 'react';

import { BrowserRouter } from 'react-router-dom';
import { render, act, cleanup, getByText, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { useSearch } from '@/features/search';
import { media } from '@/tests/factories';
import { SearchPage } from '@/pages';

describe('SearchPage', () => {
  const useSearchMock = jest.mocked(useSearch, true);

  afterEach(() => {
    cleanup();
  });

  it('renders correctly', () => {
    act(() => {
      useSearchMock.mockImplementation(() => ({
        loading: false,
        results: [],
      }));
    });

    const { asFragment } = render(<SearchPage />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('when there is a query and no results', () => {
    it('renders an error message', () => {
      act(() => {
        useSearchMock.mockImplementation(() => ({
          loading: false,
          results: [],
        }));
      });

      const { container } = render(<SearchPage />);

      act(() => {
        const searchBarTextField = container.querySelector('.search-bar-text-field');
        if (searchBarTextField) {
          userEvent.type(searchBarTextField, 'query');
        }
      });

      waitFor(() => {
        expect(getByText(container, 'No results found')).toBeDefined();
      });
    });
  });

  describe('whne there is a query and results', () => {
    it('renders with a medias grid', () => {
      act(() => {
        useSearchMock.mockImplementation(() => ({
          loading: false,
          results: [media()],
        }));
      });

      const { container } = render(<SearchPage />, { wrapper: BrowserRouter });

      act(() => {
        const searchBarTextField = container.querySelector('.search-bar-text-field');
        if (searchBarTextField) {
          userEvent.type(searchBarTextField, 'query');
        }
      });

      waitFor(() => {
        expect(getByText(container, 'No results found')).toBeUndefined();
        expect(container.querySelector('.medias-grid')).toBeDefined();
      });
    });
  });
});
