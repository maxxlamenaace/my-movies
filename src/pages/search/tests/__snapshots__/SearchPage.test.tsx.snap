// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
  />
  <div
    class="MuiBox-root css-f1zhho"
  >
    <div
      class="css-1p5q5e5-MuiStack-root"
    >
      <div
        class="css-1m9ibdo-MuiStack-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth search-bar-button css-1h77tdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          actor
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth search-bar-button css-18xp20k-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          movie
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth search-bar-button css-1h77tdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          serie
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root search-bar-text-field css-136h9lv-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            placeholder="What movie are you looking for?"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
